[ CoreData Entities ] add support for delete operations

draganescu opened the pull request:
## Description
Adds support for delete operations to entities. For a testing example use #22428

## Screenshots
![2020-04-13 15-13-04 2020-04-13 15_13_59](https://user-images.githubusercontent.com/107534/79119602-6c613680-7d99-11ea-95fb-f7ea2a78004e.gif)

## Types of changes
New feature to the entity actions and reducers that enable one to call `delete[EntityName]`

draganescu committed:
reset, delete nav menu items

there is a missing state preservation when menus are changed on the menus editor component

draganescu commented:
I've added a `snackbar` notice for removing menu items. The problem I am facing is that while in #21486 I was able to curtail the cache of the entity store by creating local state out of the inital get menus call, here the edit menu component is remounted and the menu items are taken from the cache again.

As a quick fix we could not use entities :) or otherwise we need the delete option in place. In absence, the delete works but on switching between menus the cache wrongly shows deleted menu items too.

cc @noisysocks

adamziel commented on packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js:
```diff
@@ -24,12 +27,32 @@ function createMenuItemAttributesFromBlock( block ) {
+}
+
+async function deleteMenuItem( recordId ) {
+	const path = `${ '/__experimental/menu-items/' +
+		recordId +
+		'?force=true' }`;
```
```suggestion
	const path = `/__experimental/menu-items/${ recordId }?force=true`;
```

adamziel commented on packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js:
```diff
@@ -87,10 +110,19 @@ export default function useNavigationBlocks( menuId ) {
+				{
+					type: 'snackbar',
+				}
+			);
```
Should we only have a single notification in place for all the operations? I imagine seeing "Created 10 menu items", "Deleted 2 menu items" and "Updated 3 menu items" would be pretty confusing.

draganescu replied:
Sure!

adamziel reviewed (changes requested):
I left a few small notes

noisysocks commented on packages/core-data/src/queried-data/reducer.js:
```diff
@@ -68,9 +71,9 @@ export function getMergedItemIds( itemIds, nextItemIds, page, perPage ) {
  * @return {Object} Next state.
  */
 function items( state = {}, action ) {
+	const key = action.key || DEFAULT_ENTITY_KEY;
```
We can leave this where it was if we have the `DELETE_ITEMS` action contain `itemIds` instead of `items`.

noisysocks commented on packages/core-data/src/actions.js:
```diff
@@ -139,6 +155,11 @@ export function receiveEmbedPreview( url, preview ) {
 
+export function* deleteEntityRecord( kind, name, recordId ) {
+	const edits = { toDelete: true };
+	yield editEntityRecord( kind, name, recordId, edits );
```
Is there a reason we're implementing `deleteEntityRecord` using `editEntityRecord`? It makes the already-complicated `editEntityRecord` action more complicated, and potentially will cause issues if someone tries to use `@wordpress/data` with an entity that contains a `toDelete` property 🙂

I was expecting `deleteEntityRecord` to look something like this snippet which I've spliced together from different parts of `editEntityRecord` and have not tested:

```js
export function* deleteEntityRecord( kind, name, recordId ) {
	const entities = yield getKindEntities( kind );
	const entity = find( entities, { kind, name } );
	if ( ! entity ) {
		return;
	}

	yield {
		type: 'DELETE_ENTITY_RECORD_START',
		kind,
		name,
		recordId,
	};

	const persistedRecord = yield select(
		'getEntityRecord',
		kind,
		name,
		recordId
	);

	yield removeItems( kind, name, recordId );

	let error;

	try {
		let path = `${ entity.baseURL }/${ recordId }`;

		if ( entity.forceDelete ) {
			path += '?force=true';
		}

		yield apiFetch( {
			path,
			method: 'DELETE',
		} );
	} catch ( _error ) {
		error = _error;

		yield receiveEntityRecords(
			kind,
			name,
			persistedRecord,
			undefined,
			true
		);
	}

	yield {
		type: 'DELETE_ENTITY_RECORD_FINISH',
		kind,
		name,
		recordId,
		error,
	};
}
```

draganescu replied:
My not so great reason is that I *think* that we lose undo if we go around edit. And also because my logic is that delete is just an edit until you save. In our case we're saving immediately but in other cases we would not.

I also think that the correct implementation would be to have in edit-navigation an entity provider which would hook the action that removes the block to a delete entity call and then on save we actually perform the DELETE call.

noisysocks replied:
There's some intricacies here and so, first of all, thank you for forcing me to do some research here and think through them 😀

Dispatching `undo()` and `redo()` only affects the data stored in an entity's `edits` state. In other words, only changes made to an entity using `editEntityRecord()` (viewable using `getEditedEntityRecord()`) are changed when one dispatches `undo()` or `redo()`.

`saveEntityRecord()`, `getEntityRecord()` and `deleteEntityRecord()` operate on a different level to `undo()`, `redo()` and `editEntityRecord()`. The former actions are concerned with loading, saving and deleting the entity as a whole whereas the latter actions are concerned only with the entity's `edits` state.

I wouldn't expect that one would be able to undo the changes made by `deleteEntityRecord()` just as I wouldn't expect one to be able to undo the changes made by `saveEntityRecord()` or by (the fetching that occurs in) `getEntityRecord()`.

You're right that we will want to one day implement undo and redo in `edit-navigation`. To do this we will likely stop storing the `core/navigation` and `core/navigation-link` blocks in React state as we do now and instead store them in the `edits` of a `post` entity that is never saved. That will let us use `undo()` and `redo()` to undo and redo e.g. removing a `core/navigation-link` block. When the _Save Navigation_ button is clicked, the blocks in the unsaved `post` entity will be translated into `saveEntityRecord( 'menu-item', ... )` and `deleteEntityRecord( 'menu-item', ... )` calls.

> on save we actually perform the DELETE call.

This already happens. `menu-item`s are not created or deleted until _Save Navigation_ is clicked which in turn calls `saveBlocks()` in `useNavigationBlocks()`.

noisysocks commented on packages/core-data/src/actions.js:
```diff
@@ -94,6 +94,22 @@ export function receiveEntityRecords(
+		name,
+		invalidateCache,
+	};
+}
```
This action should be in `packages/core-data/src/queried-data/actions.js` which is where `receiveItems` is. 

I also think we should name it `removeItems` / `REMOVE_ITEMS` as the action removes data from the store and doesn't perform the actual `DELETE` request.

And, as I mentioned in another thread, we only need the `itemIDs`, not all of `items`.

draganescu replied:
Agree on itemIds. On the location I figured we're not using any query, and I am not sure what is in querried-data and what is not in querried-data :)

noisysocks commented on packages/core-data/src/queried-data/reducer.js:
```diff
@@ -1,7 +1,7 @@
 /**
```
I think we may also need to update the `entity.edits` reducer in `packages/core-data/src/reducer.js` to remove any stored edits when `DELETE_ITEMS` is dispatched.

draganescu replied:
I think you are right.

draganescu commented on packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js:
```diff
@@ -14,6 +14,7 @@ function createBlockFromMenuItem( menuItem ) {
 	return createBlock( 'core/navigation-link', {
 		label: menuItem.title.raw,
 		url: menuItem.url,
+		id: menuItem.id,
```
I think this is a mistake, because I am also working on the dynamic links PR which introduces id in the block. Not sure that is the reason but good catch!

draganescu commented on packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js:
```diff
@@ -46,9 +47,11 @@ export default function useNavigationBlocks( menuId ) {
-			const block = createBlockFromMenuItem( menuItem );
-			menuItemsRef.current[ block.clientId ] = menuItem;
-			innerBlocks.push( block );
+			if ( menuItem ) {
```
Yes this was my biggest surprise as well. I am not sure and can't understand if I introduced a bug by wrongly removing from state (maybe it's b/c I am not removing edits as well?) or it's a bug in core-data. I'll see.

draganescu commented on packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js:
```diff
@@ -87,11 +90,30 @@ export default function useNavigationBlocks( menuId ) {
-			// TODO - delete menu items.
+			const menuItem = menuItemsRef.current[ clientId ];
+			deleteMenuItem( menuItem.id );
+			saveMenuItem( {
```
Because it's the save that performs the API calls. Delete is just an edit that marks the item to-delete.

noisysocks replied:
I think `deleteMenuItem` (AKA `deleteEntityRecord`) should do everything, see https://github.com/WordPress/gutenberg/pull/21557#discussion_r420547180.

adamziel commented:
@draganescu I think I got it to work, the culprit was the `invalidateCache: true` sent with the `REMOVE_ITEMS` action

draganescu committed:
fixing a bad merge

draganescu committed:
no invalidateCache and refactored according to review

draganescu committed:
revert change to getMergedItemIds

draganescu commented on docs/designers-developers/developers/data/data-core.md:
```diff
@@ -471,6 +471,10 @@ _Returns_
 
+<a name="deleteEntityRecord" href="#deleteEntityRecord">#</a> **deleteEntityRecord**
+
+Undocumented declaration.
```
This doc needs to be added

draganescu commented on docs/designers-developers/developers/data/data-core.md:
```diff
@@ -613,6 +617,10 @@ _Returns_
 
+<a name="removeItems" href="#removeItems">#</a> **removeItems**
+
+Undocumented declaration.
```
This doc needs to be added

draganescu commented on packages/core-data/src/actions.js:
```diff
@@ -94,6 +94,16 @@ export function receiveEntityRecords(
+		items: castArray( records ),
+		kind,
+		name,
+		invalidateCache: false,
```
Doesn't seem like there is a default.

draganescu committed:
refactor according to self review

noisysocks commented:
This is looking really good!

We need to add unit tests for the two new actions (`removeItems` and `deleteEntityRecord`) as well as the new branch (`REMOVE_ITEMS`) in the reducer.

Let's also add a note to `@wordpress/core-data`'s `CHANGELOG.md`.

draganescu committed:
deletes query items, adds tests for new entity methods

also adds changelog

adamziel commented:
@draganescu I rebased, tested a few scenarios, and here's what happened:

![2020-05-21 16-32-41 2020-05-21 16_36_11](https://user-images.githubusercontent.com/205419/82571111-c7e5c780-9b82-11ea-9159-fab9d97ef474.gif)

![2020-05-21 16-45-59 2020-05-21 16_48_04](https://user-images.githubusercontent.com/205419/82571158-d9c76a80-9b82-11ea-8769-f6d6b16e73f7.gif)

draganescu commented:
Apparently nested items break the deletions

noisysocks commented on packages/core-data/src/utils/on-sub-key.js:
```diff
@@ -24,8 +30,25 @@ export const onSubKey = ( actionProperty ) => ( reducer ) => (
 		return state;
 	}
 
+	if ( action.type === 'REMOVE_ITEMS' ) {
```
`onSubKey` is a generic higher order reducer which should not care about any particular action. This logic should be moved up to the reducer that uses `onSubKey`.

draganescu replied:
This is the logic that removes items from the queries branch of the state.

draganescu replied:
Yeah I got [a bit confused](https://github.com/WordPress/gutenberg/pull/22428#discussion_r428771569) but all is clear now. Will fix.

draganescu committed:
refactored for properly deleteing entity ids from querries

draganescu committed:
updates tests

draganescu committed:
moved the REMOVE_ITEMS reducer to receive all query keys

props @aduth

draganescu committed:
removes the need to send query on delete

draganescu commented:
I've updated the delete logic to support properly clearing the removed ids from queries as well ( props @aduth ) and now both branches seem to work well as far as deleting sequential items. I'll continue with:

1) a fix for nested items being wrongfully added to the delete list
2) a research into how to clean up the queriedData tree for situations like this:

<img width="378" alt="image" src="https://user-images.githubusercontent.com/107534/82691181-2b94f100-9c66-11ea-9dd1-8aff6e17c0c1.png">

Menu 224 is deleted but on the menuItems branch `menu=224` still exists as a key.

adamziel reviewed (approved):
LGTM and works. Looping in @youknowriad @gziolo @aduth who might have some additional insight about the core data part of this PR.

draganescu commented:
I'll wait for @epiqueras to take a look. Do note that @adamziel 's #22603 removes the need for this PR to manage actually deleting menu items. So the only important thing is the delete* availability for all entities.

epiqueras commented on packages/core-data/src/actions.js:
```diff
@@ -139,6 +139,70 @@ export function receiveEmbedPreview( url, preview ) {
+ *
+ * @param {string} kind      Kind of the deleted entity.
+ * @param {string} name      Name of the deleted entity.
+ * @param {Object} recordId  Record to be deleted.
```
This is a string and it's the record ID.

epiqueras commented on packages/core-data/src/actions.js:
```diff
@@ -139,6 +139,70 @@ export function receiveEmbedPreview( url, preview ) {
+		return;
+	}
+
+	yield {
+		type: 'DELETE_ENTITY_RECORD_START',
+		kind,
+		name,
+		recordId,
+	};
```
Shouldn't this be used in the reducer?

draganescu replied:
See this [comment](https://github.com/WordPress/gutenberg/pull/21557#discussion_r434035095)

epiqueras commented on packages/core-data/src/actions.js:
```diff
@@ -139,6 +139,70 @@ export function receiveEmbedPreview( url, preview ) {
+		recordId,
+	};
+
+	yield removeItems( kind, name, Number( recordId ) );
```
Why is the ID being coerced to a number?

draganescu replied:
As far as I can remember it is a string here but in the state it is a number (the state is keyed by this id as int) and in the reducer removing it didn't work.

draganescu replied:
🤦 non-number IDs!

epiqueras commented on packages/core-data/src/actions.js:
```diff
@@ -139,6 +139,70 @@ export function receiveEmbedPreview( url, preview ) {
+		);
+	}
+
+	yield {
+		type: 'DELETE_ENTITY_RECORD_FINISH',
+		kind,
+		name,
+		recordId,
+		error,
+	};
```
This is also unused.

draganescu replied:
DELETE_ENTITY_RECORD_START and DELETE_ENTITY_RECORD_FINISH are unused. They are here for consistency with the flow in `saveEntityRecord`, but can be removed. What do you think?

epiqueras commented on packages/core-data/src/actions.js:
```diff
@@ -139,6 +139,70 @@ export function receiveEmbedPreview( url, preview ) {
+			path,
+			method: 'DELETE',
+		} );
+	} catch ( _error ) {
+		error = _error;
+
+		const persistedRecord = yield select(
+			'getEntityRecord',
+			kind,
+			name,
+			recordId
+		);
+
+		yield receiveEntityRecords(
+			kind,
+			name,
+			persistedRecord,
+			undefined,
+			false
+		);
+	}
```
If it errors, we also need to restore the queries updated by `removeItems`.

I think we also should restore the edited record, not the persisted one.

draganescu replied:
> I think we also should restore the edited record, not the persisted one.

This basically means:
```
const persistedRecord = yield select(
			'getEditedEntityRecord',
			kind,
			name,
			recordId
		);
```
Right?

epiqueras commented on packages/core-data/src/queried-data/reducer.js:
```diff
@@ -134,6 +136,33 @@ const queries = flowRight( [
+			forEach( newState, ( queryItems, key ) => {
+				if ( newState[ key ] ) {
+					newState[ key ] = filter( queryItems, ( queryId ) => {
+						return ! action.items.includes( queryId );
```
We should build a POJO of `action.items` to make this lookup faster. It could really hurt performance if we have a UI with lots of pagination.

epiqueras commented on packages/core-data/src/queried-data/actions.js:
```diff
@@ -17,6 +17,25 @@ export function receiveItems( items ) {
 
+/**
+ * Returns an action object used in signalling that entity records have been
+ * deleted and it needs to be removed from entities state.
```
```suggestion
 * deleted and they need to be removed from entities state.
```

epiqueras commented on packages/core-data/src/queried-data/actions.js:
```diff
@@ -17,6 +17,25 @@ export function receiveItems( items ) {
+		items: castArray( records ),
+		kind,
+		name,
+		invalidateCache: false,
```
This needs to be a parameter, only set to true in `deleteEntityRecord`, and handled in `getEntityRecords.shouldInvalidate`.

draganescu replied:
Why is this needed? How does `getEntityRecords.shouldInvalidate` work and how does `deleteEntityRecord` impact that? Currently I am not treating this action in `getEntityRecords.shouldInvalidate` and it works. When does it break?

This line is momentarily cruft :)

epiqueras commented on packages/core-data/src/queried-data/actions.js:
```diff
@@ -17,6 +17,25 @@ export function receiveItems( items ) {
+ * Returns an action object used in signalling that entity records have been
+ * deleted and it needs to be removed from entities state.
+ *
+ * @param {string} kind            Kind of the removed entity.
```
```suggestion
 * @param {string} kind            Kind of the removed entities.
```

epiqueras commented on packages/core-data/src/queried-data/actions.js:
```diff
@@ -17,6 +17,25 @@ export function receiveItems( items ) {
+ * deleted and it needs to be removed from entities state.
+ *
+ * @param {string} kind            Kind of the removed entity.
+ * @param {string} name            Name of the removed entity.
```
```suggestion
 * @param {string} name            Name of the removed entities.
```

epiqueras commented on packages/core-data/src/queried-data/test/reducer.js:
```diff
@@ -99,4 +99,36 @@ describe( 'reducer', () => {
+		} );
+		const state = reducer( original, {
+			type: 'REMOVE_ITEMS',
+			query: [ 3 ],
```
I don't think this is needed here.

draganescu replied:
Thanks, this is a leftover from when the query reducer was implemented wrong :)

draganescu commented:
Considering #22603 solves delete as well I will remove all changes that actually delete menu items and have this be only about entity delete support.

draganescu committed:
updates some comments

also removes cruft from a test

draganescu committed:
refactoring according to review

- removes superfluous actions from deleteEntityRecords
- treats invalidateCache properly
- attempts to make lookups faster for items when removing querries

draganescu committed:
do not clear cache on delete

draganescu commented:
Howdy @epiqueras I've refactored according to review and it's ready for a new pass. I had two challenges, which I am unsure about:

- [clearing the cache ](https://github.com/WordPress/gutenberg/pull/21557/files#diff-83cecb38834613f359225842d097fc6eR158) surprisingly makes the delete now update the state so I resorted to not clearing it. 

- [this](https://github.com/WordPress/gutenberg/pull/21557/files#diff-1bd974a63140cd57a1af3c3819b742c4R165) was my understanding of your POJO suggestion, not sure if that is what you meant.

draganescu committed:
fixes test after reming superfluous actions from deleteEntityRecord

epiqueras commented:
>clearing the cache surprisingly makes the delete now update the state so I resorted to not clearing it.

You need to clear the cache after the API call that deletes the item. Otherwise, it will refetch it before it's removed.

>this was my understanding of your POJO suggestion, not sure if that is what you meant.

You need to key them by ID. Right now, you are unnecessarily calling `toPlainObject`, and then keying by an identity function, the default.

draganescu committed:
makes a proper POJO for removing items from queries and invalidates the cache after items are removed

epiqueras commented on packages/core-data/CHANGELOG.md:
```diff
@@ -2,6 +2,14 @@
 
 ## Unreleased
 
+## 2.3.1 (2019-05-21)
```
Isn't this still unreleased?

epiqueras commented on packages/core-data/CHANGELOG.md:
```diff
@@ -2,6 +2,14 @@
+### New Feature
+
+- The `deleteEntityRecord` and `removeItems` actions have been added.
+- Entities now support a `forceDelete` attribute to allow the API to delete things that do not go to trash.
```
Why isn't this an option of `deleteEntityRecord`? `deleteEntityRecord` should accept query parameters if this is needed because otherwise, we are tying the entity abstraction to the specifics of the WordPress REST API. `?force=true` might not be the way this is done for all entities.

draganescu replied:
Hm good point. To be honest it's hard to tell when you are tying to the WP api and when you're doing normal WordPress way things :)

draganescu committed:
fix changelog and add deleteQueryParams to deleteRecord

epiqueras commented on docs/designers-developers/developers/data/data-core.md:
```diff
@@ -471,6 +471,17 @@ _Returns_
+-   _kind_ `string`: Kind of the deleted entity.
+-   _name_ `string`: Name of the deleted entity.
+-   _recordId_ `string`: Record Id of the entity to be deleted.
+-   _deleteQueryParams_ `string`: Special query parameters for the DELETE API call. These parameters will be appended to the rest base with a '?' in front.
```
This should be `query` to match the other functions.

draganescu replied:
done!

draganescu committed:
fixes test

draganescu committed:
Rename deleteQueryParams to query for consistency

epiqueras commented on packages/core-data/src/actions.js:
```diff
@@ -139,6 +139,52 @@ export function receiveEmbedPreview( url, preview ) {
+ *
+ * @param {string} kind              Kind of the deleted entity.
+ * @param {string} name              Name of the deleted entity.
+ * @param {string} recordId          Record Id of the entity to be deleted.
```
```suggestion
 * @param {string} recordId          Record ID of the deleted entity.
```

epiqueras commented on packages/core-data/src/actions.js:
```diff
@@ -139,6 +139,52 @@ export function receiveEmbedPreview( url, preview ) {
+ * @param {string} kind              Kind of the deleted entity.
+ * @param {string} name              Name of the deleted entity.
+ * @param {string} recordId          Record Id of the entity to be deleted.
+ * @param {string} query Special query parameters for the DELETE API call. These parameters will be appended to the rest base with a '?' in front.
```
The query should be an object like in all the other methods. We have utilities to create the query strings.

draganescu committed:
Use the new deleteEntityRecord to delete menus (#22428)

* adds delete menu with entity delete

* updates the delete and removes the stateMenus

* passes the new force query param

* fix bug with resetting current menu after delete

draganescu committed:
fixes query param's type for consistency

epiqueras commented on packages/core-data/src/actions.js:
```diff
@@ -139,6 +144,52 @@ export function receiveEmbedPreview( url, preview ) {
+/**
+ * Action triggered to delete an entity record.
+ *
+ * @param {string} kind              Kind of the deleted entity.
+ * @param {string} name              Name of the deleted entity.
+ * @param {string} recordId          Record ID of the deleted entity.
+ * @param {?Object}query             Special query parameters for the DELETE API call.
```
The alignment is off here.

epiqueras commented on packages/core-data/src/actions.js:
```diff
@@ -139,6 +144,52 @@ export function receiveEmbedPreview( url, preview ) {
+
+		yield removeItems( kind, name, recordId, true );
+	} catch ( _error ) {
+		const persistedRecord = yield select(
+			'getEditedEntityRecord',
+			kind,
+			name,
+			recordId
+		);
+
+		yield receiveEntityRecords(
+			kind,
+			name,
+			persistedRecord,
+			undefined,
+			false
+		);
```
We don't need to do this since we are not optimistically removing items before an API response.

epiqueras commented on packages/core-data/src/queried-data/reducer.js:
```diff
@@ -138,6 +140,37 @@ const queries = flowRight( [
+		case 'REMOVE_ITEMS':
+			const newState = { ...state };
+			const removedItems = action.items.reduce( ( result, item ) => {
+				result[ item.toString() ] = true;
```
What do you expect `item.toString()` to be here?

draganescu replied:
Bah, it's just an artifact of various testing.

draganescu committed:
comment linting, removed useless catch logic for deleteEntityRecord, removed useless string conversion

draganescu committed:
try implement error handling for delete

draganescu committed:
fixed the intentional typo and the unintentional one

epiqueras commented on packages/core-data/CHANGELOG.md:
```diff
@@ -2,6 +2,11 @@
 
+### New Feature
+
+- The `deleteEntityRecord` and `removeItems` actions have been added.
```
Missing new functions.

epiqueras commented on packages/core-data/src/actions.js:
```diff
@@ -139,6 +144,58 @@ export function receiveEmbedPreview( url, preview ) {
+		error,
+	};
+
+	return deletedRecord;
```
Why do we return this?

draganescu replied:
It's an easy way to check in the caller that the delete happened. Save does the same and it is used for the same reason to not do anything if there was no save.

epiqueras commented on packages/core-data/src/queried-data/reducer.js:
```diff
@@ -138,6 +140,37 @@ const queries = flowRight( [
+		case 'REMOVE_ITEMS':
+			const newState = { ...state };
+			const removedItems = action.items.reduce( ( result, item ) => {
+				result[ item ] = true;
```
`item` is an object isn't it?

draganescu replied:
No, it's an int. `action.items` is `[2, 12, 34]`

epiqueras replied:
That doesn't match the type signature of `removeItems`.

It should be `itemIds` then.

draganescu replied:
good point.

epiqueras commented on packages/core-data/src/reducer.js:
```diff
@@ -276,6 +276,25 @@ function entity( entityConfig ) {
+									action.type ===
+									'DELETE_ENTITY_RECORD_START',
+								error: action.error,
+								isAutosave: action.isAutosave,
```
Copy-paste leftovers?

draganescu replied:
Yeesss.

epiqueras commented on packages/core-data/src/actions.js:
```diff
@@ -139,6 +144,58 @@ export function receiveEmbedPreview( url, preview ) {
+			method: 'DELETE',
+		} );
+
+		yield removeItems( kind, name, recordId, true );
```
This doesn't match the type signature of `removeItems`.

draganescu replied:
updated the description of `removeItems` signature to get `{Array|number}` for `records`

epiqueras replied:
Doesn't that mean Array or number?

draganescu replied:
that's what it means yes :)

epiqueras commented on packages/edit-navigation/src/components/menus-editor/index.js:
```diff
@@ -28,24 +37,36 @@ export default function MenusEditor( { blockEditorSettings } ) {
-		false
+	const menuDeleteError = useSelect( ( select ) =>
+		select( 'core' ).getLastEntityDeleteError( 'root', 'menu', menuId )
 	);
```
Why are there 2 `useSelect` calls?

draganescu replied:
code blindness.

epiqueras commented on packages/edit-navigation/src/components/menus-editor/index.js:
```diff
@@ -11,14 +11,23 @@ import {
 import CreateMenuArea from './create-menu-area';
 import NavigationEditor from '../navigation-editor';
 
+const noticeId = 'edit-navigation-delete-menu-error';
```
What if there are two instances of this component?

draganescu replied:
This component is the entire editor, hardly worth optimizing for two instances. The create menu, for the same reason, does not assume multiple instance.

epiqueras replied:
React itself can break this in future updates like Concurrent mode.

draganescu replied:
made it unique in the same fashion media replace flow handles this

epiqueras commented on packages/edit-navigation/src/components/menus-editor/index.js:
```diff
@@ -28,24 +37,36 @@ export default function MenusEditor( { blockEditorSettings } ) {
 			setHasCompletedFirstLoad( true );
 		}
-	}, [ hasLoadedMenus ] );
+	}, [ menus, hasLoadedMenus ] );
```
Why is this needed?

draganescu committed:
updates and fixes according to review

draganescu committed:
rename remove items' action id collection

draganescu committed:
makes notices unique in menu editor

epiqueras commented on packages/core-data/src/queried-data/actions.js:
```diff
@@ -17,6 +17,26 @@ export function receiveItems( items ) {
+ * Returns an action object used in signalling that entity records have been
+ * deleted and they need to be removed from entities state.
+ *
+ * @param {string} kind             Kind of the removed entities.
+ * @param {string} name             Name of the removed entities.
+ * @param {Array|number} records    Records removed.
```
```suggestion
 * @param {Array|number} records    Record IDs of the removed entities.
```

epiqueras replied:
Parameters are not aligned.

epiqueras commented on packages/core-data/src/queried-data/reducer.js:
```diff
@@ -86,8 +86,10 @@ function items( state = {}, action ) {
 				}, {} ),
 			};
+		case 'REMOVE_ITEMS':
+			const newState = omit( state, action.items );
```
itemIds?

epiqueras commented on packages/core-data/src/queried-data/reducer.js:
```diff
@@ -138,6 +140,37 @@ const queries = flowRight( [
+				return result;
+			}, {} );
+			forEach( newState, ( queryItems, key ) => {
+				if ( newState[ key ] ) {
```
When would `newState[ key ]` not be defined?

draganescu replied:
Given the shifts and changes in this PR no idea. Can only test and get back.

draganescu replied:
Normally this should never be undefined since all the items to delete already should be in the state. I think I ended up with some weird queries while working on deleteEntityRecord or something ... Removed it as it looks useless.

epiqueras commented on packages/core-data/src/queried-data/test/reducer.js:
```diff
@@ -113,4 +113,35 @@ describe( 'reducer', () => {
+
+	it( 'deletes an item', () => {
+		const original = deepFreeze( {
+			items: {
+				1: { id: 1, name: 'abc' },
+				2: { id: 2, name: 'def' },
+				3: { id: 3, name: 'ghi' },
+				4: { id: 4, name: 'klm' },
+			},
+			queries: {
+				'': [ 1, 2, 3, 4 ],
+				's=a': [ 1, 3 ],
+			},
+		} );
+		const state = reducer( original, {
+			type: 'REMOVE_ITEMS',
+			items: [ 3 ],
+		} );
+
+		expect( state ).toEqual( {
+			items: {
+				1: { id: 1, name: 'abc' },
+				2: { id: 2, name: 'def' },
+				4: { id: 4, name: 'klm' },
+			},
+			queries: {
+				'': [ 1, 2, 4 ],
+				's=a': [ 1 ],
```
itemIds

epiqueras commented on packages/core-data/src/queried-data/test/reducer.js:
```diff
@@ -113,4 +113,35 @@ describe( 'reducer', () => {
+			},
+		} );
+		const state = reducer( original, {
+			type: 'REMOVE_ITEMS',
```
You should use the action creator here to catch errors like the incorrect property name.

epiqueras commented on packages/core-data/src/test/actions.js:
```diff
@@ -30,6 +31,29 @@ describe( 'editEntityRecord', () => {
+		fulfillment.next();
+
+		// Start
+		fulfillment.next( entities );
```
We should probably test the whole function.

epiqueras commented on packages/edit-navigation/src/components/menus-editor/index.js:
```diff
@@ -11,41 +16,72 @@ import {
-		};
-	}, [] );
+const noticeId = uniqueId(
+	'navigation-editor/menu-editor/edit-navigation-delete-menu-error'
```
This still means that all instances of the component will share the same ID. This should be in a ref.

draganescu committed:
Update packages/core-data/src/queried-data/actions.js

Co-authored-by: Enrique Piqueras <epiqueras@users.noreply.github.com>

draganescu committed:
fixes bugs introduced by renaming items to itemsIds in the remove items action

also improves tests and made sure the notifications always have unique ids

draganescu committed:
lint

epiqueras commented on packages/edit-navigation/src/components/menus-editor/index.js:
```diff
@@ -19,33 +25,65 @@ import CreateMenuArea from './create-menu-area';
 
+	const noticeId = useRef();
+
+	noticeId.current = uniqueId(
```
This will create a new ID on every render. And you shouldn't mutate refs outside effects.

draganescu committed:
moves noticeId creation inside effect

epiqueras commented on packages/edit-navigation/src/components/menus-editor/index.js:
```diff
@@ -19,33 +25,64 @@ import CreateMenuArea from './create-menu-area';
-	const [ stateMenus, setStateMenus ] = useState();
-	const [ showCreateMenuPanel, setShowCreateMenuPanel ] = useState( false );
+	// Handle REST API Error messages.
+	useEffect( () => {
+		if ( menuDeleteError ) {
+			// Error messages from the REST API often contain HTML.
+			// createErrorNotice does not support HTML in error text, so first
+			// strip HTML out using DOMParser.
+			const document = new DOMParser().parseFromString(
+				menuDeleteError.message,
+				'text/html'
+			);
+			const errorText = document.body.textContent || '';
+			noticeId.current = uniqueId(
+				'navigation-editor/menu-editor/edit-navigation-delete-menu-error'
+			);
+			createErrorNotice( errorText, { id: noticeId.current } );
+		}
+	}, [ menuDeleteError ] );
```
Why recreate the notice ID when `menuDeleteError` changes? Will `removeNotice` always reference the right ID with this approach?

draganescu replied:
If the errors are the same they should not stack but if they're different they should. I think this is accomplished this way.

epiqueras commented on packages/core-data/src/queried-data/reducer.js:
```diff
@@ -138,6 +140,35 @@ const queries = flowRight( [
+			}, {} );
+			forEach( newState, ( queryItems, key ) => {
+				newState[ key ] = filter( queryItems, ( queryId ) => {
+					return !! removedItems[ queryId ] === false;
```
This is coercing a boolean to a boolean.

draganescu replied:
No, we're making undefined false. I am keeping (filter) everything that is not in `removedItems` and hence check for `!! removedItems[ queryId ] === false` and I am interested in when `removedItems[ queryId ]` is undefined.

epiqueras replied:
Then you could just return `removedItems[ queryId ]`.

draganescu replied:
```
const removedItems = { a: true, c: true };
const newState = [ "a", "b" ];
console.log( _.filter( newState, ( queryId ) => {
  return removedItems[ queryId ];
} ) );
=> 
["a"]
```
I want `["b"]` of that, so `! removedItems[ queryId ]`, still simpler then my expression, Thanks!

epiqueras commented on packages/core-data/src/test/actions.js:
```diff
@@ -30,6 +31,39 @@ describe( 'editEntityRecord', () => {
+			'DELETE_ENTITY_RECORD_FINISH'
+		);
+
+		expect( fulfillment.next().value ).not.toBe( false );
```
Check for `.done`.

draganescu replied:
Do you mean replace 
`expect( fulfillment.next().value ).not.toBe( false );`
with
`expect( fulfillment.next().done ).toBe( true );`
?

chrisvanpatten replied:
`.toBe()` or `.not.toBe()`… that is the question

epiqueras replied:
I would check against the whole expected shape of `.next()`.

draganescu committed:
updated according to review

- test delete for final shape when generator is done
- simplify filter removing items from query

draganescu committed:
lint
